### ISSUE TITLE
How does this compare to jax?

### ISSUE THREAD

**Issue by michaelklachko** [rsokl/MyGrad#232]

**Comment by rsokl** [comment]
MyGrad (which was written by... just me) is much simpler and is less ambitious than jax (which was written by a team at Harvard's information theory / probabilistic systems group and then was scooped by Google). 

MyGrad is a pure-numpy/python auto-grad library that seeks to "feel" just like NumPy - it was born out of an educational program that I run. That being said, it is quite efficient in its implementation.

jax provides a jit, which optimized computations and support for GPUs, which is really nice. It also has some support for higher-order derivatives. But it also less permissive about how you can write your code to be back-propped through.

Some other nice things about MyGrad:
- Its test suite is better; I use hypothesis for property-based tests to test for correctness. As a result, I have much better assurances that my ops behave / backprop through things like broadcasting and advanced indexing. IMHO you are less likely to get incorrect results from MyGrad
- Its code is easy to understand and is under the MIT license.
- It has built-in functionality for a lot of common neural network operations, like n-dimensional pooling and convolutions.
- I think MyGrad might still be the only autograd library to support backprop through completely generic `einsum` calls.

**Comment by michaelklachko** [comment]
Thank you for detailed response. Can you give an example of how "permissive" is it about my code? 

What if I want to specify some op that is not normally differentiable in Pytorch/Jax, how would I implement it using your framework? For example numpy has unpackbits op. If I want to backprop through it (ignoring that it does not make much sense to do so for a second), what do I need to do in MyGrad?

**Comment by rsokl** [comment]
Writing your own operation is pretty straightforward. Here is an implementation of `Multiply`:

```python
from mygrad import Tensor
from mygrad.operation_base import BroadcastableOp

class Multiply(BroadcastableOp):
    def __call__(self, a, b):
        """ Parameters
            ----------
            a : mygrad.Tensor
            b : mygrad.Tensor

            Returns
            ---------
            numpy.ndarray"""
        self.variables = (a, b)
        out = a.data * b.data
        return out

    def backward_var(self, grad, index, **kwargs):
        """ Parameters
            -------------
            grad : numpy.ndarray
                The incoming gradient
            index : int 
                The index of the argument being back-propped through
            Returns
            ---------
            numpy.ndarray
                The gradient backpropped through `self.variables[index]`"""
        a, b = self.variables
        if index == 0:  # backprop through a
            return grad * b.data
        elif index == 1:  # backprop through b
            return grad * a.data

def multiply(a, b, constant=False):
    """ ``f(a, b) -> a * b``

    Parameters
    ----------
    a : array_like
    b : array_like
    constant : bool, optional(default=False)
        If ``True``, the returned tensor is a constant (it
        does not back-propagate a gradient)

    Returns
    -------
    mygrad.Tensor"""
    return Tensor._op(Multiply, a, b, constant=constant)
```

**Comment by michaelklachko** [comment]
Great! This looks very similar to Pytorch, but it's easier to trace what's going on: no calls to c++/cuda code, and not too many abstraction layers. And I do like einsum! A while ago I used it to implement capsnet layer: https://github.com/michaelklachko/CapsNet/blob/master/capsnet_cifar.py#L78

Any shortcomings worth mentioning (other than the obvious lack of GPU acceleration)?

**Comment by rsokl** [comment]
> Any shortcomings worth mentioning (other than the obvious lack of GPU acceleration)?

My response to this depends on your desired application. If it is: I want "numpy with autodifferentiation", then this should cover the essentials. You can also check out the [docs](https://mygrad.readthedocs.io/en/latest/) to see if anything important is missing.

I strive to have complete parity with the most essential aspects of NumPy's semantics (e.g. broadcasting and all indexing schemes) and with the interfaces of the NumPy functions that are re-implemented by MyGrad. That being said, the more obscure `ufunc` arguments (`out`, `where`, `subok`, `casting`) are not supported, nor are the  [ufunc methods](https://numpy.org/devdocs/reference/ufuncs.html#methods) supported (e.g. `mygrad.add.at` or `mygrad.add.outer`).

Casting between dtypes mid-computation is not well fleshed out, and backpropping doesn't provide very fine control over the datatype that you want (specifically if you want to force the intermediate gradients to be represented in low precision). These are simply things I have not gotten to because they have not proven to be important for my use cases.

Overall, I think that MyGrad is a very solid NumPy + autodiff solution 😄 . Please let me know if you do find any shortcomings – I will be eager to address them. 

>  And I do like einsum! A while ago I used it to implement capsnet layer

Ooh nice! Yeah it took quite a bit of work to support backprop through arbitrary traces in `einsum`, but it is there now!
